README
slideextract
============

slideextract is a program to extract slides from videos.
It outputs an image for each detected slide. Slide detection is done by
comparing consecutive frames for differences within a specific threshold,
to account for video encoding artifacts.

For better results and faster extraction you should specify a comparison
region, such as the slide number. From a series of unchanging comparison
regions slideextract will pick the last frame in hopes of capturing the
final state of the entire slide (animated bullet points, etc...).

Once extraction is completed you can view the extracted slides with your
favorite image viewer and delete possible duplicates. This often happens
when the video jumps back and forth between slides.

You can then combine the slides to a pdf with a tool like convert:

    convert `ls -v outprefix*.png` outprefix.pdf


Dependencies
------------

* opencv 4.x


Build Instructions
------------------

    make
    make install # DESTDIR= PREFIX=/usr/local


License
-------

BSD 2-Clause license, see the COPYING file for more details.



CHANGELOG
0.3.1
=====
* Fix imwrite() parameters to work with OpenCV 4.7

0.3.0
=====
* [BREAK] Change handling of -r due to inconsistencies between docs and
  implementation.
  New API takes a point, width, and height in the format of: X.Y:WxH
* Fix extraction of final slide
* Fix handling of -h
* Add 360p test video and related SVG files

0.2.0
=====
* Add README to source distribution.
* Add GUI example to README.
* Fix region of interest handling. You can now select your ROI from any side:
  e.g. bottom-left to top-right and not just top-left to bottom-right.

0.1.0
=====
* Upgrade to OpenCV4
* Downgrade project to C++ due to OpenCV dropping the C API
* Replace autotools with simple make (GNU/BSD)

0.0.3
=====
* Quickfix opencv 2 header include issues.

0.0.2
=====
* Add man page

0.0.1
=====
* Fix incompatibilities with newer autotools versions

0.0.0
=====
* Initial release


main.c
/* slideextract -- extract slides from video
 *
 * Copyright (c) 2013, Angelo Haller <angelo@szanni.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "slideextract.h"
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdbool.h>

//! Print version and exit program
static void
version ()
{
	puts("slideextract " VERSION);

	exit (0);
}

//! Print help and exit program
static void
help()
{
	puts ("Usage: slideextract [OPTION] infile outprefix\n\
\n\
Extract slides from video.\n\
\n\
  -g          Launch gui to set comparison region, press any key when done\n\
  -r X.Y:WxH  Manually set comparison region by specifying a starting point X.Y\n\
              and width & height WxH\n\
\n\
  -h  Display this help and exit\n\
  -V  Output version information and exit\n\
\n\
Select a comparision region (e.g. slide number) for faster and more\n\
accurate extraction.\n");

	exit (0);
}

int
main (int argc, char **argv)
{
	struct roi roi;
	bool gflag = 0;
	bool rflag = 0;
	int c;
	int ret;

	while ((c = getopt(argc, argv, "gr:Vh")) != -1) {
		switch (c) {
			case 'g':
				gflag = 1;
				break;

			case 'r':
				rflag = 1;
				if (sscanf(optarg, "%d.%d:%dx%d", &roi.x, &roi.y, &roi.width, &roi.height) != 4)
					help();
				break;

			case 'V':
				version();
				break;

			case 'h':
			case '?':
			default:
				help();
		}
	}
	argc -= optind;
	argv += optind;

	if (argc != 2 || (gflag && rflag))
		help();

	const char *file = *argv++;
	const char *outprefix = *argv;

	if (rflag)
		return se_extract_slides(file, outprefix, &roi);

	if (gflag)
	{
		if ((ret = se_select_roi(file, &roi)))
			return ret;

		printf("Selected ROI: %d.%d:%dx%d\n", roi.x, roi.y, roi.width, roi.height);
		return se_extract_slides(file, outprefix, &roi);
	}

	return se_extract_slides(file, outprefix, NULL);
}



slideextract.h
/* slideextract -- api header
 *
 * Copyright (c) 2013, Angelo Haller <angelo@szanni.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef SLIDEEXTRACT_H
#define SLIDEEXTRACT_H

#ifdef __cplusplus
extern "C" {
#endif

struct roi
{
  int x;
  int y;
  int width;
  int height;
};

/*!
 * \brief Playback video for roi selection with mouse
 * \param file Video file
 * \param roi Region of interest selected by user
 * \return 0 on success, -1 on failure or abort by user
 */
int se_select_roi(const char *file, struct roi *roi);

/*!
 * \brief Extract slides from video
 * \param file Video file
 * \param outprefix Output filename prefix
 * \param roi Frame comparison region for slide detection, whole frame on NULL
 * \return 0 on success, -1 on failure
 */
int se_extract_slides(const char *file, const char *outprefix, struct roi *roi);

#ifdef __cplusplus
}
#endif

#endif



slideextract.cpp
/* slideextract -- implementation
 *
 * Copyright (c) 2013, Angelo Haller <angelo@szanni.org>
 * Copyright (c) 2022, Oleksiy Klymenko
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "slideextract.h"
#include <stdio.h>
#include <stdlib.h>

#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/videoio.hpp>

using namespace cv;

static bool selected_roi = 0;
static Point point1;
static Point point2;

static void
_se_mouseHandler(int event, int x, int y, int flags, void *param)
{
	Mat clone;
	Mat frame = *(Mat*)param;

	if (event == EVENT_LBUTTONDOWN && (flags & EVENT_FLAG_LBUTTON)) {
		point1 = Point (x, y);
		point2 = Point (x, y);
	}

	if (!(flags & EVENT_FLAG_LBUTTON))
		return;

	if (event != EVENT_MOUSEMOVE && event != EVENT_LBUTTONUP)
		return;

	point2 = Point(x, y);
	selected_roi = (point1.x != point2.x && point1.y != point2.y);

	clone = frame.clone();
	rectangle(clone, point1, point2, CV_RGB(0, 255, 0), 1, 8, 0);
	imshow("frame", clone);
}

int
se_select_roi(const char *file, struct roi *roi)
{
	VideoCapture capture = VideoCapture(file, CAP_ANY);

	if (!capture.isOpened())
		return -1;

	int key = -1;
	Mat frame;
	Mat clone;
	namedWindow("frame", WINDOW_AUTOSIZE);

	while (key == -1)
	{
		capture.read(frame);
		if (frame.empty())
			break;

		setMouseCallback("frame", _se_mouseHandler, &frame);

		if (selected_roi)
		{
			clone = frame.clone();
			rectangle(clone, point1, point2, CV_RGB(0, 255, 0), 1, 8, 0);
			imshow("frame", clone);
		}
		else
		{
			imshow("frame", frame);
		}

		key = waitKey (10);
	}

	startWindowThread();

	destroyAllWindows();

	if (selected_roi)
	{
		roi->x = (point1.x < point2.x) ? point1.x : point2.x;
		roi->y = (point1.y < point2.y) ? point1.y : point2.y;
		roi->width = abs(point2.x - point1.x);
		roi->height = abs(point2.y - point1.y);
		return 0;
	}

	return -1;
}

static double
_se_compare_image(Mat &img1, Mat &img2, struct roi *roi)
{
	double correlation;
	Mat result = Mat(1, 1, CV_32F);
	Mat i1, i2;

	if (roi != NULL) {
		Rect r = Rect(roi->x, roi->y, roi->width, roi->height);
		i1 = img1.clone()(r);
		i2 = img2.clone()(r);
	}
	else {
		i1 = img1;
		i2 = img2;
	}

	matchTemplate(i1, i2, result, TM_CCORR_NORMED);
	minMaxLoc(result, NULL, &correlation);

	return correlation;
}

int
se_extract_slides(const char *file, const char *outprefix, struct roi *roi)
{
	int num = 0;
	char str[64];

	std::vector<int> image_properties;
	image_properties.push_back(IMWRITE_PNG_COMPRESSION);
	image_properties.push_back(9);

	Mat current;
	Mat last;

	VideoCapture capture = VideoCapture(file, CAP_ANY);
	if (!capture.isOpened())
		return -1;

	bool is_final_frame = false;
	while (!is_final_frame) {
		capture.read(current);

		is_final_frame = current.empty();
		bool is_changed_frame = !current.empty() && !last.empty()
			&& _se_compare_image (last, current, roi) <= 0.999;

		if (is_final_frame || is_changed_frame) {
			snprintf (str, sizeof(str)/sizeof(*str), "%s%d.png", outprefix, num++);
			imwrite(str, last, image_properties);
		}

		last = current.clone();
	}

	return 0;
}



src/slideextract.c
/* slideextract -- implementation
 *
 * Copyright (c) 2013, Angelo Haller <angelo@szanni.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "slideextract.h"
#include <stdio.h>
#include <opencv2/core/core_c.h>
#include <opencv2/highgui/highgui_c.h>
#include <opencv2/imgproc/imgproc_c.h>
#include <opencv2/imgcodecs/imgcodecs_c.h>
#include <opencv2/videoio/videoio_c.h>

static _Bool selected_roi = 0;
static CvPoint point1;
static CvPoint point2;

static void
_se_mouseHandler (int event, int x, int y, int flags, void *param)
{
  IplImage *clone = NULL;
  IplImage *frame = param;

  if (event == CV_EVENT_LBUTTONDOWN && !(flags & CV_EVENT_FLAG_LBUTTON))
    {
      point1 = cvPoint (x, y);
      point2 = cvPoint (x, y);
    }

  if (!(flags & CV_EVENT_FLAG_LBUTTON))
    return;

  if (event != CV_EVENT_MOUSEMOVE && event != CV_EVENT_LBUTTONUP)
    return;

  point2 = cvPoint (x, y);
  selected_roi = (point1.x != point2.x && point1.y != point2.y);

  clone = cvCloneImage (frame);
  cvRectangle (clone, point1, point2, CV_RGB(0, 255, 0), 1, 8, 0);
  cvShowImage ("frame", clone);
  cvReleaseImage (&clone);
}

int
se_select_roi (const char *file, CvRect *roi)
{
  CvCapture *capture = cvCaptureFromFile (file);

  if (capture == NULL)
    return -1;

  int key = -1;
  IplImage *frame = NULL;
  IplImage *clone = NULL;
  cvNamedWindow ("frame", CV_WINDOW_AUTOSIZE);

  while (key == -1)
    {
      frame = cvQueryFrame (capture);
      if (frame == NULL)
        break;

      cvSetMouseCallback ("frame", _se_mouseHandler, frame);

      if (selected_roi)
        {
          clone = cvCloneImage (frame);
          cvRectangle (clone, point1, point2, CV_RGB(0, 255, 0), 1, 8, 0);
          cvShowImage ("frame", clone);
          cvReleaseImage (&clone);
        }
      else
        {
          cvShowImage ("frame", frame);
        }

      key = cvWaitKey (10);
    }

  cvDestroyWindow ("frame");
  cvReleaseCapture (&capture);
  cvStartWindowThread (); //memory leak

  if (selected_roi)
    {
      *roi = cvRect(point1.x, point1.y, point2.x - point1.x, point2.y - point1.y);
      return 0;
    }

  return -1;
}

static double
_se_compare_image (IplImage *img1, IplImage *img2, CvRect *roi)
{
  double correlation;
  CvMat *result = cvCreateMat (1, 1, CV_32F);

  if (roi != NULL)
    {
      cvSetImageROI (img1, *roi);
      cvSetImageROI (img2, *roi);
    }

  cvMatchTemplate (img1, img2, result, CV_TM_CCORR_NORMED);
  cvMinMaxLoc (result, NULL, &correlation, NULL, NULL, NULL);
  cvReleaseMat (&result);

  if (roi != NULL)
    {
      cvResetImageROI (img1);
      cvResetImageROI (img2);
    }

  return correlation;
}

int
se_extract_slides (const char *file, const char *outprefix, CvRect *roi)
{
  int num = 0;
  char str[64];
  int image_properties[3] = {CV_IMWRITE_PNG_COMPRESSION, 9, 0};
  IplImage *current = NULL;
  IplImage *last = NULL;
  CvCapture *capture = cvCaptureFromFile (file);

  if (capture == NULL)
    return -1;

  do
    {
      current = cvQueryFrame (capture);
      if (current == NULL)
        break;

      if (last != NULL)
        {
          double cmp = _se_compare_image (last, current, roi);
          if (cmp <= 0.999)
            {
              snprintf (str, sizeof(str)/sizeof(*str), "%s%d.png", outprefix, num++);
              cvSaveImage (str, last, image_properties);
            }
          cvReleaseImage (&last);
        }
      last = cvCloneImage (current);
    }
  while (current != NULL);

  cvReleaseImage (&last);
  cvReleaseCapture (&capture);

  return 0;
}

